### PR TITLE
SIP 2 - Create Credential Manifests CRUD

### PR DESCRIPTION
**Pull request by nitro-neal**
This is the implementation of the Credential Manifest CRUD functionality according to SIP 2
https://github.com/TBD54566975/ssi-service/tree/main/sip/sips/sip_2

image

**Postman Data**d
PUT Manifest

Input: 
```
{
  "@context": "",
  "issuer" : "did:example:123",
  "outputDescriptors": [
            {
                "id": "id1z1",
                "schema": "https://test.com/schema",
                "name": "good ID",
                "description": "it's all good"
            },
            {
                "id": "id2z",
                "schema": "https://test.com/schema",
                "name": "good ID",
                "description": "it's all good"
            }
        ],
   "presentationDefinition": {
    "id": "test",
    "input_descriptors": [
      {
        "id": "test input descriptor",
        "constraints" : 
            {
                "fields" : [ 
                    {
                    "path" : [
                        "path1"
                    ]
                    }
                ]
            }
        

      }
    ]
  }
}
```

Output
```
{
    "manifest": {
        "id": "a2170fed-81cd-42db-ad79-3d65ea720878",
        "issuer": {
            "id": "did:example:123",
            "name": "did:example:123"
        },
        "output_descriptors": [
            {
                "id": "id1z1",
                "schema": "https://test.com/schema",
                "name": "good ID",
                "description": "it's all good"
            },
            {
                "id": "id2z",
                "schema": "https://test.com/schema",
                "name": "good ID",
                "description": "it's all good"
            }
        ],
        "format": {
            "jwt": {
                "alg": [
                    "EdDSA"
                ]
            }
        },
        "presentation_definition": {
            "id": "test",
            "input_descriptors": [
                {
                    "id": "test input descriptor",
                    "constraints": {
                        "fields": [
                            {
                                "path": [
                                    "path1"
                                ]
                            }
                        ]
                    }
                }
            ]
        }
    }
}
```

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -24,6 +24,8 @@ require (
 
 require go.etcd.io/bbolt v1.3.6
 
+require github.com/oliveagle/jsonpath v0.0.0-20180606110733-2e52cf6e6852 // indirect
```
Comment (nitro-neal) [inline go.mod:27]: I don't see where I added this or used this? When I try to remove it brings it back when I mage build

File: pkg/server/router/manifest.go [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+// @Param        issuer   query     string  false  "string issuer"
+// @Param        schema   query     string  false  "string schema"
+// @Param        subject  query     string  false  "string subject"
+// @Success      200      {object}  GetManifestsResponse
+// @Failure      400      {string}  string  "Bad request"
+// @Failure      500      {string}  string  "Internal server error"
+// @Router       /v1/manifests [get]
+func (cr ManifestRouter) GetManifests(ctx context.Context, w http.ResponseWriter, r *http.Request) error {
+	gotManifests, err := cr.service.GetManifests()
+
```
Comment (decentralgabe) [inline pkg/server/router/manifest.go:143]: where do you handle the query parameters?

File: pkg/server/router/manifest.go [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+// @Summary      Get manifest
+// @Description  Get manifest by id
+// @Tags         ManifestAPI
+// @Accept       json
+// @Produce      json
+// @Param        id   path      string  true  "ID"
+// @Success      200  {object}  GetManifestResponse
+// @Failure      400  {string}  string  "Bad request"
+// @Router       /v1/manifests/{id} [get]
+func (cr ManifestRouter) GetManifest(ctx context.Context, w http.ResponseWriter, r *http.Request) error {
```
Comment (decentralgabe) [inline pkg/server/router/manifest.go:102]: nit: for schemas it's called `GetSchemaByID`, I think your name here is better, would you mind updating the schema to say `GetSchema` ?

File: pkg/service/manifest/storage/bolt.go [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+		errMsg := fmt.Sprintf("could not unmarshal stored manifest: %s", id)
+		logrus.WithError(err).Error(errMsg)
+		return nil, errors.Wrapf(err, errMsg)
+	}
+	return &stored, nil
+}
+
+// GetManifests attempts to get all stored manifests. It will return those it can even if it has trouble with some.
+func (b BoltManifestStorage) GetManifests() ([]StoredManifest, error) {
+	gotmanifests, err := b.db.ReadAll(namespace)
```
Comment (decentralgabe) [inline pkg/service/manifest/storage/bolt.go:66]: gotManifests

File: pkg/service/manifest/storage/bolt.go [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+		logrus.WithError(err).Error(errMsg)
+		return nil, errors.Wrap(err, errMsg)
+	}
+	if len(gotmanifests) == 0 {
+		logrus.Info("no manifests to get")
+		return nil, nil
+	}
+	var stored []StoredManifest
+	for _, manifestBytes := range gotmanifests {
+		var nextmanifest StoredManifest
```
Comment (decentralgabe) [inline pkg/service/manifest/storage/bolt.go:78]: nextManifest

File: pkg/service/manifest/model.go [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+import (
+	manifestsdk "github.com/TBD54566975/ssi-sdk/credential/manifest"
+)
+
+type CreateManifestRequest struct {
+	Issuer string
+	// A context is optional. If not present, we'll apply default, required context values.
+	Context                string
+	OutputDescriptors      []map[string]interface{}
+	PresentationDefinition map[string]interface{}
```
Comment (decentralgabe) [inline pkg/service/manifest/model.go:12]: these should use the object models from the sdk

**Review by decentralgabe — CHANGES_REQUESTED** [review]
one more thing - the `config.toml` files should be updated to have a new manifest entry

File: pkg/service/manifest/storage/bolt.go [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+func NewBoltManifestStorage(db *storage.BoltDB) (*BoltManifestStorage, error) {
+	if db == nil {
+		return nil, errors.New("bolt db reference is nil")
+	}
+	return &BoltManifestStorage{db: db}, nil
+}
+
+func (b BoltManifestStorage) StoreManifest(manifest StoredManifest) error {
+	//id := manifest.Schema.ID
```
Comment (decentralgabe) [inline pkg/service/manifest/storage/bolt.go:27]: remove

**Review by decentralgabe — APPROVED** [review]
Past the change to `config.toml`, lgtm!

**Pull request merged**